arrow-continuations/src/main/kotlin/arrow/continuations/generic/DelimContScope.kt

```diff
@@ -84,42 +82,113 @@ class DelimContScope<R>(val f: suspend DelimitedScope<R>.() -> R) : DelimitedSco
     DelimContScope(f).invoke()
 
   @Suppress("UNCHECKED_CAST")
-  fun invoke(): R {
-    f.startCoroutineUninterceptedOrReturn(this, Continuation(EmptyCoroutineContext) { result ->
-      resultVar.value = result.getOrThrow()
-    }).let {
-      if (it == COROUTINE_SUSPENDED) {
-        // we have a call to shift so we must start execution the blocks there
-        resultVar.loop { mRes ->
-          if (mRes === EMPTY_VALUE) {
-            val nextShiftFn = nextShift.getAndSet(null)
-              ?: throw IllegalStateException("No further work to do but also no result!")
-            nextShiftFn.startCoroutineUninterceptedOrReturn(Continuation(EmptyCoroutineContext) { result ->
-              resultVar.value = result.getOrThrow()
-            }).let { nextRes ->
-              // If we suspended here we can just continue to loop because we should now have a new function to run
-              // If we did not suspend we short-circuited and are thus done with looping
-              if (nextRes != COROUTINE_SUSPENDED) resultVar.value = nextRes as R
+  suspend fun invoke(): R {
+    val result = f.startCoroutineUninterceptedOrReturn(this, Continuation(coroutineContext) { result ->
+      promise.setResult(result.getOrThrow())
+    })
+    // We suspended, we might receive a result or encounter shiftFn
+    if (result == COROUTINE_SUSPENDED) {
+      while (true) { // Loop as long as we encounter `shift` functions, stop looping when we find an `R`
+        val nextShiftOrResult = promise.await()
+        when (val shiftOrNull = nextShiftOrResult as? (suspend () -> R)) {
```

Comment:
Pray that no-one will ever return `suspend () -> R` from a computation block :trollface: